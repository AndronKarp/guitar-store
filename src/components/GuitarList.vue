<template>
  <b-container fluid class="mt-3">
    <b-row cols-xl="4" cols-lg="3" cols-md="2" cols="1">
      <GuitarListItem
        v-for="(guitar, index) in guitars"
        :key="index"
        :guitar="guitar"
      />
    </b-row>
  </b-container>
</template>

<script>
import GuitarListItem from "./GuitarListItem";
import { mapGetters } from "vuex";

export default {
  computed: {
    ...mapGetters(["guitars"])
  },
  components: {
    GuitarListItem
  }
};
</script>

<style></style>
