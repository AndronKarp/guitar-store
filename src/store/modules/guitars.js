export default {
  state: {
    guitars: [],
    areGuitarsFetched: false
  },
  getters: {
    guitars(state) {
      return state.guitars;
    },
    areGuitarsFetched(state) {
      return state.areGuitarsFetched;
    }
  },
  mutations: {
    pushToGuitars(state, guitar) {
      state.guitars.push(guitar);
    },
    setNewGuitarQuantity(state, { guitar, value }) {
      guitar.quantity = value;
    },
    updateAreGuitarsFetchedStatus(state, { status }) {
      state.areGuitarsFetched = status;
    }
  },
  actions: {
    addToGuitars(store, guitar) {
      store.commit("pushToGuitars", guitar);
    },
    updateGuitarQuantity(context, { guitar, value }) {
      context.commit("setNewGuitarQuantity", { guitar, value });
    },
    setAreGuitarsFetchedStatusToTrue(store) {
      store.commit("updateAreGuitarsFetchedStatus", { status: true });
    }
  }
};
