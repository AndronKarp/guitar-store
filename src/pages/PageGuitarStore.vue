<template>
  <div
    class="d-flex flex-fill"
    :class="{ 'justify-content-center': !areGuitarsFetched }"
  >
    <GuitarList v-if="areGuitarsFetched" />
    <b-spinner
      v-else
      class="align-self-center"
      label="Loading..."
      variant="info"
    ></b-spinner>
  </div>
</template>

<script>
import GuitarList from "../components/GuitarList";
import { mapGetters } from "vuex";

export default {
  computed: {
    ...mapGetters(["areGuitarsFetched"])
  },
  components: {
    GuitarList
  }
};
</script>
<style></style>>
