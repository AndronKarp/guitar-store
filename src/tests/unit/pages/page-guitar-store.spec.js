import { shallowMount, createLocalVue } from "@vue/test-utils";
import PageGuitarStore from "@/pages/PageGuitarStore";
import GuitarList from "@/components/GuitarList";
import Vuex from "vuex";
import guitars from "@/store/modules/guitars";
import BootstrapVue from "bootstrap-vue";

const localVue = createLocalVue();

localVue.use(BootstrapVue);
localVue.use(Vuex);

describe("PageGuitarStore.vue", () => {
  let state;
  let store;
  let wrapper;

  beforeEach(() => {
    state = { areGuitarsFetched: true };
    store = new Vuex.Store({
      state,
      getters: guitars.getters
    });
    wrapper = shallowMount(PageGuitarStore, {
      localVue,
      store
    });
  });

  test("renders guitar store after guitars were fetched", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("renders loading spinner while guitars are fetching", () => {
    state.areGuitarsFetched = false;
    wrapper.vm.$nextTick(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  test("accepts GuitarList child component", () => {
    const guitarList = wrapper.findComponent(GuitarList);
    expect(guitarList.exists()).toBe(true);
  });
});
