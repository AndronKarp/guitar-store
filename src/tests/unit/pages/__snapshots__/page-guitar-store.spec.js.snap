// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageGuitarStore.vue renders guitar store after guitars were fetched 1`] = `
<div class="d-flex w-100 h-100">
  <guitarlist-stub></guitarlist-stub>
</div>
`;

exports[`PageGuitarStore.vue renders loading spinner while guitars are fetching 1`] = `
<div class="d-flex w-100 h-100 justify-content-center">
  <b-spinner-stub type="border" label="Loading..." variant="info" role="status" tag="span" class="align-self-center"></b-spinner-stub>
</div>
`;
